### PR TITLE
API: added refactor of genetic code

### PR DESCRIPTION
**Pull request by GavinHuttley**

### CODE REVIEW

File: src/cogent3/core/new_genetic_code.py [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+from cogent3.util.table import Table
+
+
+OptStr = typing.Optional[str]
+SetStr = typing.Set[str]
+ConverterType = typing.Callable[[bytes, bytes], bytes]
+StrORInt = typing.Union[str, int]
+
+DNA = new_moltype.get_moltype("dna")
+RNA = new_moltype.get_moltype("rna")
```
Comment (fredjaya) [inline src/cogent3/core/new_genetic_code.py:30]: Can't place if these are used in the file? e.g. could be used on L95 `moltype: new_moltype.MolType = DNA`, but having the reference to the `new_*` module in-line is clearer.

File: src/cogent3/core/new_genetic_code.py [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+    def start_codons(self) -> SetStr:
+        return self._start_codons
+
+    @property
+    def sense_codons(self) -> SetStr:
+        return self._sense_codons
+
+    def to_table(self):
+        """returns aa to codon mapping as a cogent3 Table"""
+        from cogent3.core.moltype import IUPAC_PROTEIN_code_aa
```
Comment (fredjaya) [inline src/cogent3/core/new_genetic_code.py:143]: any reason why import isn't placed at the top of the file? Also refer to `moltype.IUPAC_PROTEIN_code_aa` in-line?

File: src/cogent3/core/new_genetic_code.py [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+
+        rows = []
+        headers = ["aa", "IUPAC code", "codons"]
+        for code, aa in IUPAC_PROTEIN_code_aa.items():
```
Comment (fredjaya) [inline src/cogent3/core/new_genetic_code.py:147]: `moltype.IUPAC_PROTEIN_code_aa`

File: src/cogent3/core/new_genetic_code.py (hunk 2 above)
Comment (KatherineCaley) [reply src/cogent3/core/new_genetic_code.py:143]: should be importing from new_moltype?

File: src/cogent3/core/new_genetic_code.py [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+            return self._aa_to_codon.get(item, set())
+
+        if len(item) != 3:
+            raise InvalidCodonError(f"Codon or aa {item} has wrong length")
+
+        key = item.upper()
+        key = key.replace("U", "T")
+        return self._codon_to_aa.get(key, "X")
+
+    def translate(self, dna, start: int = 0, rc: bool = False) -> str:
```
Comment (KatherineCaley) [inline src/cogent3/core/new_genetic_code.py:185]: type hint for dna?

File: src/cogent3/core/new_genetic_code.py (hunk 1 above)
Comment (GavinHuttley) [reply src/cogent3/core/new_genetic_code.py:30]: good point

File: src/cogent3/core/new_genetic_code.py (hunk 2 above)
Comment (GavinHuttley) [reply src/cogent3/core/new_genetic_code.py:143]: Doh!!

**Pull request merged**